samples/unpeelingOnion/src/onion/ecommerce/controllers/shoppingCartController.ts

```diff
@@ -0,0 +1,160 @@
+import { NextFunction, Request, Response, Router } from 'express';
+import { CommandBus } from '#core/commands';
+import { sendCreated } from '#core/http';
+import { mongoObjectId } from '#core/mongodb';
+import { QueryBus } from '#core/queries';
+import { OpenShoppingCart } from 'src/onion/ecommerce/application/shoppingCarts/commands/shoppingCarts/openShoppingCart';
+import { GetShoppingCartById } from 'src/onion/ecommerce/application/shoppingCarts/queries/getShoppingCartById';
+import { assertNotEmptyString, assertPositiveNumber } from '#core/validation';
+import { AddProductItemToShoppingCartRequest } from 'src/onion/ecommerce/requests/shoppingCarts/addProductItemToShoppingCartRequest';
+import { AddProductItemToShoppingCart } from 'src/onion/ecommerce/application/shoppingCarts/commands/shoppingCarts/addProductItemToShoppingCart';
+import { ProductItem } from 'src/onion/ecommerce/common/shoppingCarts/productItem';
+import { RemoveProductItemFromShoppingCart } from 'src/onion/ecommerce/application/shoppingCarts/commands/shoppingCarts/removeProductItemFromShoppingCart';
+import { RemoveProductItemFromShoppingCartRequest } from 'src/onion/ecommerce/requests/shoppingCarts/removeProductItemFromShoppingCartRequest.ts';
+import { ConfirmShoppingCart } from 'src/onion/ecommerce/application/shoppingCarts/commands/shoppingCarts/confirmShoppingCart';
+import { GetCustomerShoppingHistory } from 'src/onion/ecommerce/application/shoppingCarts/queries/getCustomerShoppingHistory';
+
+export class ShoppingCartController {
+  public router = Router();
+
+  constructor(private commandBus: CommandBus, private queryBus: QueryBus) {
+    this.router.post('/customers/:customerId/shopping-carts/', this.open);
```

Comment:
You should extract /customers/:id/shopping-carts as a common router. Using router.use

Reply:
I don’t like hidden base paths. I think that keeping it explicit shows the full intention. You don’t need to Kim through multiple places to know the final URL. DRY is not always end goal, but more a tool for me.